dvc/api.py

```diff
@@ -1,15 +1,43 @@
+from contextlib import contextmanager
+import os
+import shutil
+import tempfile
+
+from dvc.utils.compat import urlparse
 from dvc.repo import Repo
+from dvc.external_repo import ExternalRepo
 
 
-def get_url(path, repo_dir=None, remote=None):
-    """Returns an url of `path` in default or specified remote"""
-    repo = Repo(repo_dir)
-    out, = repo.find_outs_by_path(path)
-    remote_obj = repo.cloud.get_remote(remote)
-    return str(remote_obj.checksum_to_path_info(out.checksum))
+def get_url(path, repo=None, remote=None):
+    """Returns an url of a resource specified by path in repo"""
+    with _make_repo(repo) as _repo:
+        abspath = os.path.join(_repo.root_dir, path)
+        out, = _repo.find_outs_by_path(abspath)
+        remote_obj = _repo.cloud.get_remote(remote)
+        return str(remote_obj.checksum_to_path_info(out.checksum))
 
 
-def open(path, repo_dir=None, remote=None, mode="r", encoding=None):
+@contextmanager
+def open(path, repo=None, remote=None, mode="r", encoding=None):
     """Opens a specified resource as a file descriptor"""
-    repo = Repo(repo_dir)
-    return repo.open(path, remote=remote, mode=mode, encoding=encoding)
+    with _make_repo(repo) as _repo:
+        abspath = os.path.join(_repo.root_dir, path)
+        with _repo.open(
+            abspath, remote=remote, mode=mode, encoding=encoding
+        ) as fd:
+            yield fd
+
+
+@contextmanager
+def _make_repo(repo_url):
+    if not repo_url or urlparse(repo_url).scheme == "":
+        yield Repo(repo_url)
+    else:
+        tmp_dir = tempfile.mkdtemp("dvc-repo")
+        try:
+            ext_repo = ExternalRepo(tmp_dir, url=repo_url)
+            ext_repo.install()
+            yield ext_repo.repo
+        finally:
+            print("Removing temp repo")
+            shutil.rmtree(tmp_dir)
```

Comment:
Please use `remove` from `dvc.utils`. It is able to handle read-only files ,that git sometimes leaves behind.